common/network-common/src/main/java/org/apache/spark/network/util/NettyUtils.java

```diff
@@ -95,6 +111,38 @@ public static String getRemoteAddress(Channel channel) {
     return "<unknown remote>";
   }
 
+  /**
+   * Returns the default number of threads for both the Netty client and server thread pools.
+   * If numUsableCores is 0, we will use Runtime get an approximate number of available cores.
+   */
+  public static int defaultNumThreads(int numUsableCores) {
+    final int availableCores;
+    if (numUsableCores > 0) {
+      availableCores = numUsableCores;
+    } else {
+      availableCores = Runtime.getRuntime().availableProcessors();
+    }
+    return Math.min(availableCores, MAX_DEFAULT_NETTY_THREADS);
+  }
+
+  /**
+   * Returns the lazily created shared pooled ByteBuf allocator for the specified allowCache
+   * parameter value.
+   */
+  public static synchronized PooledByteBufAllocator getSharedPooledByteBufAllocator(
```

Comment:
Maybe use double-checked locking instead of method synchronization, since the instantiation just needs to happen once but this may unnecessarily block all later calls.

Reply:
I have read once this article (so I try to avoid using it):
https://www.javaworld.com/article/2074979/java-concurrency/double-checked-locking--clever--but-broken.html

Reply:
What that article talks about was fixed in Java 1.5. Double-checked locking works fine since then.

But I think that's overkill here.

Reply:
If you will use volatile, this code must use two `volatile` instance variables instead of one array. This is because `volatile PooledByteBufAllocator _s[]` is effective only for `_s` rather than `_s[0], _s[1], ...`.

Reply:
Thanks! I decided to keep the `synchronized`.